### PR TITLE
AKI-655 Added multi-layer parent block query in the blockstore

### PR DESCRIPTION
**Pull request by AionJayT**
## Description

-

Fixes Issue # .

## Type of change

- [ ] Bug fix.
- [ ] New feature.
- [ ] Enhancement.
- [ ] Unit test.
- [ ] Breaking change (a fix or feature that causes existing functionality to not work as expected).
- [ ] Requires documentation update.

## Testing

-

### CODE REVIEW

File: modAionImpl/src/org/aion/zero/impl/db/AionBlockStore.java [hunk 1]
```diff
@@ -1384,6 +1384,131 @@ void redoIndexWithoutSideChains(Block block) {
...
+     */
+    public final Block[] getTwoGenParentsByHashWithInfo(byte[] hash) {
+        if (hash == null) {
+            return null;
+        }
+
+        lock.lock();
+
+        try {
+            Block block = blocks.get(hash);
```
Comment (AionJayT) [inline modAionImpl/src/org/aion/zero/impl/db/AionBlockStore.java]: make inner method call.

File: modAionImpl/src/org/aion/zero/impl/blockchain/AionBlockchainImpl.java [hunk 2]
```diff
@@ -1589,6 +1593,14 @@ private Block getParent(BlockHeader header) {
         return getBlockStore().getBlockByHashWithInfo(header.getParentHash());
     }
 
+    private Block[] getTwoGenParents(BlockHeader header) {
```
Comment (AionJayT) [inline modAionImpl/src/org/aion/zero/impl/blockchain/AionBlockchainImpl.java]: check this code!

Comment (AionJayT) [reply modAionImpl/src/org/aion/zero/impl/blockchain/AionBlockchainImpl.java]: just pass the parentHash!

File: modAionImpl/src/org/aion/zero/impl/blockchain/AionBlockchainImpl.java [hunk 3]
```diff
@@ -1597,62 +1609,58 @@ public boolean isValid(BlockHeader header) {
         if (!this.headerValidator.validate(header, LOG)) {
             return false;
         }
+        /*
+         * threeGenParents[0] = parentBlock, threeGenParents[1] = grandParent, threeGenParents[2] greatGrandParent.
+         * if threeGenParents is not equal to null, the threeGenParents[0] must have block data.
+         */
+        Block[] threeGenParents = getThreeGenParents(header);
```
Comment (AionJayT) [inline modAionImpl/src/org/aion/zero/impl/blockchain/AionBlockchainImpl.java]: create temporarily variable.

File: modAionImpl/src/org/aion/zero/impl/blockchain/AionBlockchainImpl.java [hunk 4]
```diff
@@ -1598,61 +1601,58 @@ public boolean isValid(BlockHeader header) {
...
-        if (parent == null) {
+        Block[] threeGenParents = getBlockStore().getThreeGenerationBlocksByHashWithInfo(header.getParentHash());
+
+        if (threeGenParents == null) {
             return false;
         }
 
-        Block grandParent = getParent(parent.getHeader());
+        Block parentBlock = threeGenParents[0];
+        Block greatParentBlock = threeGenParents[1];
```
Comment (AlexandraRoatis) [inline modAionImpl/src/org/aion/zero/impl/blockchain/AionBlockchainImpl.java]: minor: typo in names, should have been `grandparentBlock` & `greatGrandparentBlock`

File: modAionImpl/src/org/aion/zero/impl/db/AionBlockStore.java [hunk 5]
```diff
@@ -573,25 +573,34 @@ public Block getBlockByHashWithInfo(byte[] hash) {
...
+    /**
+     * Retrieve block with unity protocol info for the inner call
+     * @param hash given hash of the block
+     * @return the block data has matched hash with unity protocol info
+     */
+    private Block getBlockWithInfo(byte[] hash) {
+        Block retBlock = blocks.get(hash);
+        if (retBlock == null) {
+            return null;
+        }
```
Comment (AlexandraRoatis) [inline modAionImpl/src/org/aion/zero/impl/db/AionBlockStore.java]: I would prefer and `else` statement here; makes the code easier to maintain

**Pull request merged**